cleverhans/experimental/certification/utils.py

```diff
@@ -95,6 +95,67 @@ def initialize_dual(neural_net_params_object, init_dual_file=None,
               'lambda_quad': lambda_quad, 'lambda_lu': lambda_lu, 'nu': nu}
   return dual_var
 
+def lanczos_decomp(vector_prod_fn, scalar, n, k):
+  """Function that performs the Lanczos algorithm on a matrix.
+
+  Args:
+    vector_prod_fn: function which returns product H*x, where H is a matrix for
+      which we computing eigenvector.
+    scalar: quantity to scale the basis vector by (either 0 or max magnitude
+      eigenvalue)
+    n: dimensionality of matrix H
+    k: number of iterations and dimensionality of the tridiagonal matrix to
+      return
+
+  Returns:
+    alpha: vector of diagonal elements of T
+    beta: vector of off-diagonal elements of T
+    Q: orthonormal basis matrix for the Krylov subspace
+  """
+  Q = tf.zeros([n, 1])
+  v = tf.random_uniform([n, 1])
+  v = v / tf.norm(v)
+  Q = tf.concat([Q, v], axis=1)
+
+  # diagonals of the tridiagonal matrix
+  beta = tf.reshape(tf.constant(0, dtype=tf.float32), [1,])
+  alpha = tf.reshape(tf.constant(0, dtype=tf.float32), [1,])
+
+  for i in range(k):
+    v = vector_prod_fn(tf.reshape(Q[:, i+1], [n, 1])) - tf.scalar_mul(scalar, tf.reshape(Q[:, i+1], [n, 1]))
+    v = tf.reshape(v, [n,])
+    curr_alpha = tf.reshape(tf.reduce_sum(tf.multiply(v, Q[:, i+1])), [1,])
+    alpha = tf.concat([alpha, curr_alpha], axis=0)
+    v = v-beta[-1]*Q[:, i]-alpha[-1]*Q[:, i+1]
+    curr_beta = tf.reshape(tf.norm(v), [1,])
+    beta = tf.concat([beta, curr_beta], axis=0)
+    curr_norm = tf.reshape(v/(beta[-1]+1e-8), [n, 1])
+    Q = tf.concat([Q, curr_norm], axis=1)
+
+  alpha = tf.slice(alpha, begin=[1], size=[-1])
+  return alpha, beta, Q
```

Comment:
The first element of alpha, beta and Q is zero.
You remove if from alpha vector in previous line, however for beta and Q it's getting removed in the code which calls this function.
From the point of view of designing a good API it would be much better to remove the first element in this function rather than in the code which calls it.